packages/core/src/services/chains/run.ts

```diff
@@ -96,106 +95,120 @@ async function iterate({
   workspace: Workspace
   source: LogSources
   chain: Chain
-  apikeys: CachedApiKeys
+  providersMap: CachedApiKeys
   controller: ReadableStreamDefaultController
   previousCount?: number
-  previousApiKey?: ProviderApiKey
   documentLogUuid: string
-  previousResponse?: ChainTextResponse
+  previousResponse?: ChainStepResponse<StreamType>
   configOverrides?: ConfigOverrides
 }) {
-  try {
-    const step = await computeStepData({
-      chain,
-      previousResponse,
-      apikeys,
-      apiKey: previousApiKey,
-      sentCount: previousCount,
-    })
+  const prevText = previousResponse?.text
+  const stepValidator = new ChainValidator({ prevText, chain, providersMap })
+  const step = await stepValidator.call().then((r) => r.unwrap())
```

Comment:
For now we keep failing as before. Only difference is that all the errors that come from validator are already categorized and ready to be saved in `run_errors` table. 

I'm thinking on handling the saving on the `.catch` in the top promise